File: src/util/uiwatchdog.cpp
#include "uiwatchdog.h"

#include <QCoreApplication>
#include <QDateTime>
#include <QLoggingCategory>
#include <QMetaObject>

Q_LOGGING_CATEGORY(lcWatchdog, "octave.ui_watchdog")

UiWatchdog::UiWatchdog(QObject *parent, int pingMs, int stallMs)
    : QObject(parent), m_pingMs(pingMs), m_stallMs(stallMs)
{
}

UiWatchdog::~UiWatchdog()
{
    stop();
}

void UiWatchdog::start()
{
    if (m_running.exchange(true))
        return;
    m_lastPong = QDateTime::currentMSecsSinceEpoch();
    // The worker owns nothing Qt-side; it only posts queued calls to `this`,
    // which lives on the GUI thread.
    QObject::connect(&m_thread, &QThread::started, [this] { loop(); });
    m_thread.setObjectName(QStringLiteral("ui-watchdog"));
    m_thread.start();
}

void UiWatchdog::stop()
{
    if (!m_running.exchange(false))
        return;
    m_thread.quit();
    m_thread.wait(2000);
}

void UiWatchdog::loop()
{
    bool stalled = false;
    qint64 stallStart = 0;
    while (m_running.load()) {
        QThread::msleep(static_cast<unsigned long>(m_pingMs));
        if (!m_running.load())
            break;
        // Ping: runs on the GUI thread when the event loop gets to it
        QMetaObject::invokeMethod(this, [this] { m_lastPong = QDateTime::currentMSecsSinceEpoch(); },
                                  Qt::QueuedConnection);
        const qint64 now = QDateTime::currentMSecsSinceEpoch();
        const qint64 silent = now - m_lastPong.load();
        if (!stalled && silent > m_stallMs) {
            stalled = true;
            stallStart = m_lastPong.load();
            qCWarning(lcWatchdog) << "UI thread blocked for" << silent / 1000.0 << "s and counting";
        } else if (stalled && silent <= m_pingMs) {
            stalled = false;
            qCWarning(lcWatchdog) << "UI thread responsive again after"
                                  << (m_lastPong.load() - stallStart) / 1000.0 << "s";
        }
    }
}

File: backend/ui_watchdog.py
"""Detects a hung Qt event loop and says so in the log. A worker thread
pings the GUI thread every ping_s; if the GUI thread has not answered within
stall_s a WARNING is logged (once per stall, with the duration when it
recovers), so a frozen head unit leaves a trace. Mirrors
src/util/uiwatchdog.{h,cpp}."""

import threading
import time

from PySide6.QtCore import QObject, Signal, Slot, Qt

from backend.logging_config import get_logger

logger = get_logger(__name__)


class UiWatchdog(QObject):
    _ping = Signal()   # worker -> GUI thread (queued)

    def __init__(self, parent=None, ping_s: float = 2.0, stall_s: float = 5.0):
        super().__init__(parent)
        self._ping_s = ping_s
        self._stall_s = stall_s
        self._last_pong = time.monotonic()
        self._running = False
        self._thread = None
        self._ping.connect(self._pong, Qt.ConnectionType.QueuedConnection)

    @Slot()
    def _pong(self):
        # Runs on the GUI thread when the event loop gets to it
        self._last_pong = time.monotonic()

    def start(self):
        if self._running:
            return
        self._running = True
        self._last_pong = time.monotonic()
        self._thread = threading.Thread(target=self._loop, name="ui-watchdog", daemon=True)
        self._thread.start()

    def stop(self):
        self._running = False
        t = self._thread
        if t is not None and t is not threading.current_thread():
            t.join(timeout=self._ping_s + 0.5)
        self._thread = None

    def _loop(self):
        stalled = False
        stall_start = 0.0
        while self._running:
            time.sleep(self._ping_s)
            if not self._running:
                break
            try:
                self._ping.emit()
            except RuntimeError:
                # The QObject was torn down under us (shutdown without stop())
                break
            silent = time.monotonic() - self._last_pong
            if not stalled and silent > self._stall_s:
                stalled = True
                stall_start = self._last_pong
                logger.warning(f"UI thread blocked for {silent:.1f} s and counting")
            elif stalled and silent <= self._ping_s:
                stalled = False
                logger.warning(f"UI thread responsive again after {self._last_pong - stall_start:.1f} s")
